httpbp/middlewares.go

```diff
@@ -299,3 +306,103 @@ func recoverPanic(name string, next HandlerFunc) HandlerFunc {
 		return next(ctx, w, r)
 	}
 }
+
+// statusCodeRecorder is used by RecordStatusCode to record the code passed to
+// a call to WriteHeader.
+type statusCodeRecorder struct {
+	http.ResponseWriter
+
+	code int
+}
+
+func (r *statusCodeRecorder) WriteHeader(code int) {
+	r.ResponseWriter.WriteHeader(code)
+	if r.code == 0 {
+		r.code = code
+	}
+}
+
+var families = [10]string{
+	"nan",
+	"1xx",
+	"2xx",
+	"3xx",
+	"4xx",
+	"5xx",
+	"6xx",
+	"7xx",
+	"8xx",
+	"9xx",
+}
+
+// statusCodeFamily takes an http status code and returns it as an "Nxx"
+// string. Returns "nan" if code < 100 or code > 999.
+func statusCodeFamily(code int) string {
+	family := code / 100
+	if family < 0 || family >= len(families) {
+		return families[0]
+	}
+	return families[family]
+}
+
+// counterGenerator is used by recordStatusCode to create counters for recording
+// http response codes set by the server.
+//
+// this was added purely to make it easier to test the middleware.
+type counterGenerator interface {
+	Counter(name string) metrics.Counter
+}
+
+func recordStatusCode(counters counterGenerator) Middleware {
+	return func(name string, next HandlerFunc) HandlerFunc {
+		counter := counters.Counter("baseplate.http." + name + ".response")
+		return func(ctx context.Context, w http.ResponseWriter, r *http.Request) (err error) {
+			wrapped := &statusCodeRecorder{ResponseWriter: w}
+			defer func() {
+				var code int
+				if wrapped.code != 0 {
+					// WriteHeader was called explicitly, use that
+					code = wrapped.code
+				} else if err != nil {
+					// something went wrong, check if err is an HTTPErr where we can extract
+					// the code, otherwise assume InternalServerError
+					var httpErr HTTPError
+					if errors.As(err, &httpErr) {
+						code = httpErr.Response().Code
+					} else {
+						code = http.StatusInternalServerError
+					}
+				} else {
+					// if there's no error returned and no call to WriteHeader, Go will
+					// return OK.
+					// https://pkg.go.dev/net/http#ResponseWriter.WriteHeader
+					code = http.StatusOK
+				}
```

Comment:
maybe we could move this logic from ln362-380 into a method on `statusCodeRecorder`, that way I can reuse in the prometheus middleware like so: https://github.com/reddit/baseplate.go/pull/451/files#diff-f0f0f7b278ddd73593a069e51e2d62c8b1fcac7d296258c2e724d85d31c8bb9eR306